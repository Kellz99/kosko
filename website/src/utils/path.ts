export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
