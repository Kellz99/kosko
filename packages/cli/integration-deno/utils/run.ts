import execa from "execa";
import { join } from "node:path";
import { projectRoot } from "@kosko/test-utils";

export async function runDenoCLI(
  args: string[],
  options: execa.Options = {}
): Promise<execa.ExecaReturnValue> {
  return execa(
    "deno",
    [
      "run",
      "-A",
      "--import-map",
      join(projectRoot, "packages/cli/templates/deno/import_map.json"),
      "npm:kosko",
      ...args
    ],
    {
      ...options,
      env: {
        LC_ALL: "en_US",
        ...options.env
      }
    }
  );
}
