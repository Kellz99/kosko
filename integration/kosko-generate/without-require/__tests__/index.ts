import execa from "execa";
import { dirname, join } from "path";
import { runCLI } from "../../../run";
import symlinkDir from "symlink-dir";
import pkgDir from "pkg-dir";

const testDir = dirname(__dirname);

let args: string[];
let result: execa.ExecaReturnValue;
let options: execa.Options;

beforeAll(async () => {
  const root = await pkgDir();

  await symlinkDir(
    join(root!, "packages", "env"),
    join(testDir, "node_modules", "@kosko", "env")
  );
});

beforeEach(async () => {
  result = await runCLI(args, {
    ...options,
    cwd: testDir
  });
});

describe("when output is not set", () => {
  beforeAll(() => {
    args = ["generate", "--env", "dev"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should output YAML", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when output = yaml", () => {
  beforeAll(() => {
    args = ["generate", "--env", "dev", "--output", "yaml"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should output YAML", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when output = json", () => {
  beforeAll(() => {
    args = ["generate", "--env", "dev", "--output", "json"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should output JSON", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when output is invalid", () => {
  beforeAll(() => {
    args = ["generate", "--env", "dev", "--output", "foo"];
    options = { reject: false };
  });

  test("should return status code 1", () => {
    expect(result.exitCode).toEqual(1);
  });

  test("should print the error", () => {
    expect(result.stderr).toMatchSnapshot();
  });
});
