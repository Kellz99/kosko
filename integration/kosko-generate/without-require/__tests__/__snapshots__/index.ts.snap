// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when output = json should output JSON 1`] = `
"{
  \\"apiVersion\\": \\"v1\\",
  \\"kind\\": \\"List\\",
  \\"items\\": [
    {
      \\"metadata\\": {
        \\"name\\": \\"nginx\\"
      },
      \\"spec\\": {
        \\"replicas\\": 1,
        \\"selector\\": {
          \\"matchLabels\\": {
            \\"app\\": \\"nginx\\"
          }
        },
        \\"template\\": {
          \\"metadata\\": {
            \\"labels\\": {
              \\"app\\": \\"nginx\\"
            }
          },
          \\"spec\\": {
            \\"containers\\": [
              {
                \\"name\\": \\"nginx\\",
                \\"image\\": \\"nginx\\"
              }
            ]
          }
        }
      },
      \\"apiVersion\\": \\"apps/v1\\",
      \\"kind\\": \\"Deployment\\"
    },
    {
      \\"metadata\\": {
        \\"name\\": \\"nginx\\"
      },
      \\"spec\\": {
        \\"selector\\": {
          \\"app\\": \\"nginx\\"
        },
        \\"ports\\": [
          {
            \\"port\\": 80
          }
        ]
      },
      \\"apiVersion\\": \\"v1\\",
      \\"kind\\": \\"Service\\"
    }
  ]
}"
`;

exports[`when output = yaml should output YAML 1`] = `
"---
metadata:
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx
apiVersion: apps/v1
kind: Deployment
---
metadata:
  name: nginx
spec:
  selector:
    app: nginx
  ports:
    - port: 80
apiVersion: v1
kind: Service"
`;

exports[`when output is invalid should print the error 1`] = `
"kosko generate [components..]

Generate Kubernetes manifests

Positionals:
  components  Components to generate. This overrides components set in config
              file.                                        [array] [default: []]

Global Options:
  --cwd      Path of working directory                   [string] [default: CWD]
  --silent   Disable log output                       [boolean] [default: false]
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

Options:
  --env, -e      Environment name                                       [string]
  --require, -r  Require modules. Modules set in config file will also be
                 required.                                 [array] [default: []]
  --output, -o   Output format
                            [string] [choices: \\"yaml\\", \\"json\\"] [default: \\"yaml\\"]
  --validate     Validate components                   [boolean] [default: true]

Examples:
  kosko generate                      Generate manifests
  kosko generate foo bar              Specify components
  kosko generate foo_*                Use glob pattern
  kosko generate --env foo            Set environment
  kosko generate -r ts-node/register  Require external modules

Invalid values:
  Argument: output, Given: \\"foo\\", Choices: \\"yaml\\", \\"json\\""
`;

exports[`when output is not set should output YAML 1`] = `
"---
metadata:
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx
apiVersion: apps/v1
kind: Deployment
---
metadata:
  name: nginx
spec:
  selector:
    app: nginx
  ports:
    - port: 80
apiVersion: v1
kind: Service"
`;
